cpp/src/arrow/compute/api_vector.h

```diff
@@ -522,6 +533,16 @@ Result<Datum> DictionaryEncode(
     const DictionaryEncodeOptions& options = DictionaryEncodeOptions::Defaults(),
     ExecContext* ctx = NULLPTR);
 
+ARROW_EXPORT
+Result<std::shared_ptr<Array>> CumulativeSum(const Array& values,
+                                             CumulativeSumOptions& options,
+                                             ExecContext* ctx = NULLPTR);
```

Comment:
These function signatures differ from the others.
```c++
Result<Datum> CumulativeSum(const Datum& data, const CumulativeSumOptions& options = CumulativeSumOptions::Defaults(), ExecContext* ctx = NULLPTR);
```
CumulativeSumOptions& options = default options

Reply:
They are similar to those of `NthToIndices` and `SelectKUnstable`, though the default options would do the same as the default constructor of `CumulativeSumOptions`.

Reply:
IIRC we refactored `FunctionOptions` to follow this convention. Is this correct @lidavidm @pitrou?

Reply:
I don't recall if/when we did this, but I would agree that we should at least have `const Options&` (to follow the style guide), and providing default options is more convenient.